draft-ietf-httpapi-ratelimit-headers.md

```diff
@@ -937,6 +937,28 @@ there's a high probability that all clients will show up at `18:00:00`.
 
 This could be mitigated adding some jitter to the field-value.
 
+A service returning `RateLimit` fields MUST NOT convey values
+exposing an unwanted volume of requests
+and SHOULD implement mechanisms to cap the ratio between `RateLimit-Remaining`
+`RateLimit-Reset`.
+
+The following example describes a service
+with a policy of 10000 quota-units per 1000 seconds
+that have not been consumed.
+
+~~~ example
+RateLimit-Limit: 10000, 10000;w=1000
+RateLimit-Remaining: 10000
+RateLimit-Reset: 10
+~~~
+
+In the last 10 seconds, the above fields expose
+an average throughput of 1000 quota-units per second.
```

Comment:
Since we don't enforce an interpretation of the data to infer a rate, an agent might as well decide to consume most of the quota during the first second while still abiding by this data, thus having the potential to overwhelm the service if it hasn't taken into account this possibility and does not provide updated information in follow up responses.

Although a service can adapt this information as that rate grows and quota diminishes, we might want to consider adding text to say that the default interpretation for the maximum rate (ie. unless specified otherwise via policy comments, further extensions or document amendments) should be the average quota-units/s inferred from the remaining and reset values.

Reply:
> an agent might as well decide to consume most of the quota during the first second

The right thing to do is to implement multiple quotas with a shorter timeframe.

> adding text to say that the default interpretation for the maximum rate [..]
> should be the average quota-units/s inferred from the ~~remaining and reset values~~

do you mean  the average value expressed in the quota-policy?

Reply:
I mean that if no other information (like another quota policy parameter) is provided, a client could just interpret that within any given period of time, the maximum rate at which it should be consuming the service is:

`maximum-rate = expiring-limit / time-window` units/second
or
`maximum-rate = remaining / reset` units/second
or
`maximum-rate = expiring-limit / reset` units/second
or
`undefined`

(this takes into account that some of these parameters might be missing from the response, including missing the whole field)

We can discuss whether this maximum rate defined above should be defined in this particular way or some other way, but the main issue I want to raise is:

**Does it make sense to add a sensible _default_ expected maximum rate of consumption within a given time window?**

I think it is good to define what the expected maximum rate of consumption would be by default as an informative hint, to avoid the case in which we have plenty of remaining quota to consume and the client wants to take it all in a fraction of the time window resulting in a huge traffic peak and potential waste of resources.

Of course, given the huge traffic peak the service can adjust the field values in successive responses to let the client know they are going way over the expected rate before refusing requests, but this would be reactive rather than proactive.

For example, we could have a service to fetch files over HTTP, and given 100 remaining units in which each unit represents a file fetching request, the client could:

* Decide it will fetch one file at a time consuming 1 unit.
* Decide it will spawn 5 threads to fetch 5 files in parallel, consuming 5 units at the same time.
* Decide it will maximize parallelism and spawn 100 threads each one fetching a different file, consuming 100 units at the same time.

Even if the service could handle 100 units for the nominal time window period, it could very well only support 5 files to be fetched in parallel, and if so it would have to adjust the headers in the "maximum parallelism" case once it noticed the extra rate, forcing connections to wind down and threads to be terminated. Preventing this waste of resources (in both client, server, intermediaries, network, CPUs, memory, ...) could be achieved by assuming a reasonable default maximum rate limit or adding this information explicitly in the fields.

If a service could support a higher rate limit than this default interpretation then it could adjust the fields to use smaller time windows and/or bigger nominal values (just the same as if it tried to lower that default maximum rate), so this adaptive property could be preserved yet we'd have a starting point that would be "reasonable".

Reply:
> Does it make sense to add a sensible default expected maximum rate of consumption within a given time window?

I think the general rule is to carefullly sizing windows and try to follow the average number of requests per second

```
average-rate = expiring-limit / time-window units/second
```

We could even add two more informative parameters (`c: concurrency max`, `m: rps max`) but this will make the framework more complex for the general case. I would address that in another PR.

Reply:
Note that since `time-window` might be missing I used multiple ways to compute the rate. I also used the term "maximum rate" but it might be better to use "maximum average rate" or some other related concept such as maximum "surge" or "concurrency", but I guess we intuitively agree on what this represents.

> I think the general rule is to carefullly sizing windows and try to follow the average number of requests per second

That's fine, but this requires dynamically adjusting these windows, perhaps starting with conservative values to prevent traffic spikes. A quota policy parameter would most definitely resolve this issue, but since that would be optional anyway, clients might be left with no option other than figuring out some workable value (or none at all), leaving the service no option but to adapt to the client request rate. I think we can do better by just defining the expected value for this concept given the data provided by the service.

This would avoid the service being too conservative from the start, and it would help reduce the need for dynamic adaptations of time windows and rate limiting information (ie. it would help services be more effective with simple rate limiting logic that would not adapt closely to clients rates but instead would just close/limit the service altogether for a client on unsustainable traffic spikes).

> I would address that in another PR.

Sounds good! Adding parameters is something I think would provide value, because you might be able to tolerate higher traffic spikes than some definition in the text, but looks secondary to me in relation to this definition/guidance which serves as a way to ensure the default behavior is reasonable.